**Import necessary libraries:**


In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix




Loading the dataset

In [ ]:
# Load the dataset
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/wdbc.data"
columns = ['ID', 'Diagnosis'] + [f'Feature_{i}' for i in range(1, 31)]
data = pd.read_csv(url, header=None, names=columns)


To get an initial look at the dataset after preprocessing, we use the head() function. This command shows the first 10 rows, allowing us to verify that the data has been correctly processed and structured.

In [ ]:
# Show the first 5 rows
print(data.head())


         ID Diagnosis  Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  \
0    842302         M      17.99      10.38     122.80     1001.0    0.11840   
1    842517         M      20.57      17.77     132.90     1326.0    0.08474   
2  84300903         M      19.69      21.25     130.00     1203.0    0.10960   
3  84348301         M      11.42      20.38      77.58      386.1    0.14250   
4  84358402         M      20.29      14.34     135.10     1297.0    0.10030   

   Feature_6  Feature_7  Feature_8  ...  Feature_21  Feature_22  Feature_23  \
0    0.27760     0.3001    0.14710  ...       25.38       17.33      184.60   
1    0.07864     0.0869    0.07017  ...       24.99       23.41      158.80   
2    0.15990     0.1974    0.12790  ...       23.57       25.53      152.50   
3    0.28390     0.2414    0.10520  ...       14.91       26.50       98.87   
4    0.13280     0.1980    0.10430  ...       22.54       16.67      152.20   

   Feature_24  Feature_25  Feature_26  Featu

Displaying the First 10 Rows of the Dataset

In [ ]:
print(data.head(10))


         ID Diagnosis  Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  \
0    842302         M      17.99      10.38     122.80     1001.0    0.11840   
1    842517         M      20.57      17.77     132.90     1326.0    0.08474   
2  84300903         M      19.69      21.25     130.00     1203.0    0.10960   
3  84348301         M      11.42      20.38      77.58      386.1    0.14250   
4  84358402         M      20.29      14.34     135.10     1297.0    0.10030   
5    843786         M      12.45      15.70      82.57      477.1    0.12780   
6    844359         M      18.25      19.98     119.60     1040.0    0.09463   
7  84458202         M      13.71      20.83      90.20      577.9    0.11890   
8    844981         M      13.00      21.82      87.50      519.8    0.12730   
9  84501001         M      12.46      24.04      83.97      475.9    0.11860   

   Feature_6  Feature_7  Feature_8  ...  Feature_21  Feature_22  Feature_23  \
0    0.27760    0.30010    0.14710  ... 

Displaying the Last 10 Rows of the Dataset

In [ ]:
print(data.tail(10))


         ID Diagnosis  Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  \
559  925291         B      11.51      23.93      74.52      403.5    0.09261   
560  925292         B      14.05      27.15      91.38      600.4    0.09929   
561  925311         B      11.20      29.37      70.67      386.0    0.07449   
562  925622         M      15.22      30.62     103.40      716.9    0.10480   
563  926125         M      20.92      25.09     143.00     1347.0    0.10990   
564  926424         M      21.56      22.39     142.00     1479.0    0.11100   
565  926682         M      20.13      28.25     131.20     1261.0    0.09780   
566  926954         M      16.60      28.08     108.30      858.1    0.08455   
567  927241         M      20.60      29.33     140.10     1265.0    0.11780   
568   92751         B       7.76      24.54      47.92      181.0    0.05263   

     Feature_6  Feature_7  Feature_8  ...  Feature_21  Feature_22  Feature_23  \
559    0.10210    0.11120    0.04105  

The shape attribute is used to quickly view the dimensions of the dataset. It returns the number of rows and columns in the DataFrame, which helps verify the size of the data after any preprocessing steps.

In [ ]:
data.shape


(569, 32)

In [ ]:
data.isnull().sum

<bound method DataFrame.sum of         ID  Diagnosis  Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  \
0    False      False      False      False      False      False      False   
1    False      False      False      False      False      False      False   
2    False      False      False      False      False      False      False   
3    False      False      False      False      False      False      False   
4    False      False      False      False      False      False      False   
..     ...        ...        ...        ...        ...        ...        ...   
564  False      False      False      False      False      False      False   
565  False      False      False      False      False      False      False   
566  False      False      False      False      False      False      False   
567  False      False      False      False      False      False      False   
568  False      False      False      False      False      False      False   

     Feature_6  Feature_7  Feature_8  ...  Feature_21  Feature_22  Feature_23  \
0        False      False      False  ...       False       False       False   
1        False      False      False  ...       False       False       False   
2        False      False      False  ...       False       False       False   
3        False      False      False  ...       False       False       False   
4        False      False      False  ...       False       False       False   
..         ...        ...        ...  ...         ...         ...         ...   
564      False      False      False  ...       False       False       False   
565      False      False      False  ...       False       False       False   
566      False      False      False  ...       False       False       False   
567      False      False      False  ...       False       False       False   
568      False      False      False  ...       False       False       False   

     Feature_24  Feature_25  Feature_26  Feature_27  Feature_28  Feature_29  \
0         False       False       False       False       False       False   
1         False       False       False       False       False       False   
2         False       False       False       False       False       False   
3         False       False       False       False       False       False   
4         False       False       False       False       False       False   
..          ...         ...         ...         ...         ...         ...   
564       False       False       False       False       False       False   
565       False       False       False       False       False       False   
566       False       False       False       False       False       False   
567       False       False       False       False       False       False   
568       False       False       False       False       False       False   

     Feature_30  
0         False  
1         False  
2         False  
3         False  
4         False  
..          ...  
564       False  
565       False  
566       False  
567       False  
568       False  

[569 rows x 32 columns]>

Handling Column Removal and Data Mapping

In [ ]:
x=data.iloc[:,:-1]
y=data.iloc[:,-1]
print(x.shape,y.shape)

(569, 30) (569,)


Data Preparation: The dataset is split into features (x) with 30 attributes and the target variable (y) for tumor diagnosis (malignant or benign), both containing 569 samples.

In [ ]:
if 'ID' in data.columns:
    data.drop('ID', axis=1, inplace=True)

data['Diagnosis'] = data['Diagnosis'].map({'M': 1, 'B': 0})
print(data.head())


   Diagnosis  Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  \
0          1      17.99      10.38     122.80     1001.0    0.11840   
1          1      20.57      17.77     132.90     1326.0    0.08474   
2          1      19.69      21.25     130.00     1203.0    0.10960   
3          1      11.42      20.38      77.58      386.1    0.14250   
4          1      20.29      14.34     135.10     1297.0    0.10030   

   Feature_6  Feature_7  Feature_8  Feature_9  ...  Feature_21  Feature_22  \
0    0.27760     0.3001    0.14710     0.2419  ...       25.38       17.33   
1    0.07864     0.0869    0.07017     0.1812  ...       24.99       23.41   
2    0.15990     0.1974    0.12790     0.2069  ...       23.57       25.53   
3    0.28390     0.2414    0.10520     0.2597  ...       14.91       26.50   
4    0.13280     0.1980    0.10430     0.1809  ...       22.54       16.67   

   Feature_23  Feature_24  Feature_25  Feature_26  Feature_27  Feature_28  \
0      184.60      2019.0  

Getting an Overview of the Dataset

In [ ]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Diagnosis   569 non-null    int64  
 1   Feature_1   569 non-null    float64
 2   Feature_2   569 non-null    float64
 3   Feature_3   569 non-null    float64
 4   Feature_4   569 non-null    float64
 5   Feature_5   569 non-null    float64
 6   Feature_6   569 non-null    float64
 7   Feature_7   569 non-null    float64
 8   Feature_8   569 non-null    float64
 9   Feature_9   569 non-null    float64
 10  Feature_10  569 non-null    float64
 11  Feature_11  569 non-null    float64
 12  Feature_12  569 non-null    float64
 13  Feature_13  569 non-null    float64
 14  Feature_14  569 non-null    float64
 15  Feature_15  569 non-null    float64
 16  Feature_16  569 non-null    float64
 17  Feature_17  569 non-null    float64
 18  Feature_18  569 non-null    float64
 19  Feature_19  569 non-null    f

The describe() method provides summary statistics for numerical columns in the dataset

In [ ]:
data.describe()


Diagnosis   Feature_1   Feature_2   Feature_3    Feature_4  \
count  569.000000  569.000000  569.000000  569.000000   569.000000   
mean     0.372583   14.127292   19.289649   91.969033   654.889104   
std      0.483918    3.524049    4.301036   24.298981   351.914129   
min      0.000000    6.981000    9.710000   43.790000   143.500000   
25%      0.000000   11.700000   16.170000   75.170000   420.300000   
50%      0.000000   13.370000   18.840000   86.240000   551.100000   
75%      1.000000   15.780000   21.800000  104.100000   782.700000   
max      1.000000   28.110000   39.280000  188.500000  2501.000000   

        Feature_5   Feature_6   Feature_7   Feature_8   Feature_9  ...  \
count  569.000000  569.000000  569.000000  569.000000  569.000000  ...   
mean     0.096360    0.104341    0.088799    0.048919    0.181162  ...   
std      0.014064    0.052813    0.079720    0.038803    0.027414  ...   
min      0.052630    0.019380    0.000000    0.000000    0.106000  ...   
25%      0.086370    0.064920    0.029560    0.020310    0.161900  ...   
50%      0.095870    0.092630    0.061540    0.033500    0.179200  ...   
75%      0.105300    0.130400    0.130700    0.074000    0.195700  ...   
max      0.163400    0.345400    0.426800    0.201200    0.304000  ...   

       Feature_21  Feature_22  Feature_23   Feature_24  Feature_25  \
count  569.000000  569.000000  569.000000   569.000000  569.000000   
mean    16.269190   25.677223  107.261213   880.583128    0.132369   
std      4.833242    6.146258   33.602542   569.356993    0.022832   
min      7.930000   12.020000   50.410000   185.200000    0.071170   
25%     13.010000   21.080000   84.110000   515.300000    0.116600   
50%     14.970000   25.410000   97.660000   686.500000    0.131300   
75%     18.790000   29.720000  125.400000  1084.000000    0.146000   
max     36.040000   49.540000  251.200000  4254.000000    0.222600   

       Feature_26  Feature_27  Feature_28  Feature_29  Feature_30  
count  569.000000  569.000000  569.000000  569.000000  569.000000  
mean     0.254265    0.272188    0.114606    0.290076    0.083946  
std      0.157336    0.208624    0.065732    0.061867    0.018061  
min      0.027290    0.000000    0.000000    0.156500    0.055040  
25%      0.147200    0.114500    0.064930    0.250400    0.071460  
50%      0.211900    0.226700    0.099930    0.282200    0.080040  
75%      0.339100    0.382900    0.161400    0.317900    0.092080  
max      1.058000    1.252000    0.291000    0.663800    0.207500  

[8 rows x 31 columns]

In [ ]:
#Count unique values for each column
unique_counts = data.nunique()

# Display the unique counts
print(unique_counts)

Diagnosis       2
Feature_1     456
Feature_2     479
Feature_3     522
Feature_4     539
Feature_5     474
Feature_6     537
Feature_7     537
Feature_8     542
Feature_9     432
Feature_10    499
Feature_11    540
Feature_12    519
Feature_13    533
Feature_14    528
Feature_15    547
Feature_16    541
Feature_17    533
Feature_18    507
Feature_19    498
Feature_20    545
Feature_21    457
Feature_22    511
Feature_23    514
Feature_24    544
Feature_25    411
Feature_26    529
Feature_27    539
Feature_28    492
Feature_29    500
Feature_30    535
dtype: int64


Calculate the total number of rows in the dataset